persistence/javadsl/src/main/scala/com/lightbend/lagom/internal/javadsl/persistence/ReadSideActor.scala

```diff
@@ -122,4 +146,29 @@ private[lagom] class ReadSideActor[Event <: AggregateEvent[Event]](
       // This actor will be restarted by WorkerCoordinator
       throw cause
   }
+
+  private def userFlowWrapper(
+      workerCoordinates: WorkerCoordinates,
+      userFlow: Flow[japi.Pair[Event, LagomOffset], Done, _]
+  ): Flow[(japi.Pair[Event, LagomOffset], AkkaOffset), AkkaOffset, _] =
+    Flow.fromGraph(GraphDSL.create(userFlow) { implicit builder => wrappedFlow =>
+      import GraphDSL.Implicits._
+      val unzip = builder.add(Unzip[japi.Pair[Event, LagomOffset], AkkaOffset])
+      val zip   = builder.add(Zip[Done, AkkaOffset])
+      val metricsReporter: FlowShape[(Done, AkkaOffset), AkkaOffset] = builder.add(Flow.fromFunction {
+        e: (Done, AkkaOffset) =>
+          // TODO: in ReadSide processor we can't report `afterUserFlow` and `completedProcessing` separately
+          //  as we do in TopicProducerActor, unless we moved the invocation of `afterUserFlow` to each
+          //  particular ReadSideHandler (C* and JDBC).
+          ProjectionSpi.afterUserFlow(workerCoordinates.projectionName, e._2)
+          ProjectionSpi.completedProcessing(Future(e._2), context.dispatcher)
+          e._2
+      })
+
+      unzip.out0 ~> wrappedFlow ~> zip.in0
+      unzip.out1 ~> zip.in1
+      zip.out ~> metricsReporter.in
```

Comment:
This is the main change in ReadSide processors. The user-provided flow (which is user code inside a `ReadSideHandler` so both user statements and OffsetStore statments) is now wrapped and a side-channel for the offset to propagate alongside the userflow exists. The side-channel zips back and produces an `Offset` for the metrics.